### ISSUE TITLE
Measuring the significance of different benchmarking approaches with A/A testing

### ISSUE THREAD

**Issue by mdboom** [faster-cpython/ideas#480]
## Motivation

We currently do our canonical benchmarking on a single machine that we control. This has a lot of advantages: 

- Its configuration doesn't change, so we can reuse baseline measurements
- We can limit (within reason) other activities on the machine that interfere with measurement

However, for good security reasons, we can't share this resource with others in the community, especially when the benchmarking machine is on a corporate network. We therefore have to run benchmarks on others' behalf and it is difficult for others to reproduce our results independently.

Inspired by the paper [Performance Testing in the Cloud. How Base is it Really?](https://peerj.com/preprints/3507/), I decided to experiment with running benchmarks on the standard Github Actions infrastructure. This would be something much easier to share---anyone with triage rights to CPython could add a label to run benchmarks---but obviously unpredictable, multi-tenant machines introduce some challenges.

In this report `head` refers to a revision with experimental change being introduced, and `base` is the revision being compared against.

## Adapting benchmark running to CI

The key lessons of the above paper are:

- You never know what type of machine you will get or the load it is under, so `base` measurements can not be reused, but must be rerun alongside the `head` measurements every time.
- Performance within a run tends to increase or decrease as a step function due to the otherwise unobservable work of other tenants on the same machine. Some redundant measurements and post-processing to remove the outliers is required to account for this.

Therefore, we need to change the normal order of operations for bare metal to adapt it to the cloud.  The existing bare metal order of operations is:

```
- Compile head
- Run benchmark A on head
- Run benchmark B on head
- Run benchmark C on head
```

The `base` measurements were made beforehand, and can just be reused with each new `head`.  For the cloud, we need to do:

```
- Compile base
- Compile head
- Run benchmark A on base
- Run benchmark A on head
- Run benchmark B on base
- Run benchmark B on head
- Run benchmark C on base
- Run benchmark C on head
```

So clearly, more raw CPU work needs to be done for the cloud situation. However, the compilation of each branch and the running of each benchmark type can be parallelized, and through the magic of parallelism, the actual wallclock time (or time-to-answer) is less.  For example, running the pyperformance suite in Github Actions takes 36 minutes and on our benchmarking machine it takes 53 minutes.

There is an [example pull request with the CI configuration](https://github.com/mdboom/cpython/pull/1) and [an example run](https://github.com/mdboom/cpython/actions/runs/3236283530/attempts/1).

## An A/A test

To measure the variability of the results in the cloud, I performed an A/A test where the same source revision was compiled twice, and the benchmarks were run on both of these ostensibly identical builds (the reasons for not reusing an identical binary are elaborated below).  If the benchmarking system and other sources of variability were not an issue, one would expect an exact 1x performance change.  As a baseline of the "best case scenario" under ideal conditions, the same A/A test was run on the bare metal benchmarking infrastructure.

In both cases, individual outlier samples within each benchmark were removed when they are more than 2 standard deviations from the mean before being plotted in these distributions.

![plot](https://user-images.githubusercontent.com/38294/195456184-9a9c694c-a173-4130-84ba-ed4ac1a22442.png)

Unsurprisingly, it is clear that the variability of measurements is much higher on cloud than bare metal. Anecdotally, this seems to be worse on benchmarks that perform I/O (`pathlib`).

Some of the benchmarks clearly have a multi-modal distribution, such as `pidigits`.  Plotting the individual samples, you can see, perhaps, the "step function" caused by other work on the physical machine mentioned in the paper:

![scatter](https://user-images.githubusercontent.com/38294/195457126-2131b90d-b007-44a8-8fa0-84e7b86a37ed.png)

However, the majority seem to follow a normal distribution, even when they have a much wider variability than bare metal, and even when I/O isn't an issue.  For example, `go`:

![scatter-go](https://user-images.githubusercontent.com/38294/195457449-b2ef551d-0324-4db4-b75c-b6a2f68331a6.png)

I don't see an obvious way to remove signal from the noise there.

## Conclusion

Aggregating the results across benchmarks has its own set of unresolved statistical issues, but ignoring that for now -- if you aggregate and plot the 90th percentile on the distribution (the dotted line), you can get a sense of how reliable the different approaches are for decision-making.

**Bare metal**

![quantiles-bare](https://user-images.githubusercontent.com/38294/195458376-84218454-7ae1-48c6-93bc-ced3d2d62837.png)

**Cloud**

![quantiles-ci](https://user-images.githubusercontent.com/38294/195458409-bc049b7f-3b57-40a5-aaf0-132eb338306b.png)

On bare metal, a performance improvement needs to be at least 2.5% for it to be an improvement 90% of the time.  On cloud, the improvement must be 5% for the same level of certainty.

Given that we struggle to find single changes of this magnitude, bare metal isn't good enough for many things, and the cloud approach is much worse, short of some statistical silver bullet to counteract its shortcomings.  Maybe it could be useful as a first pass, since it's faster-to-answer and no-human-in-the-loop, but I think it would be rare that it alone would be conclusive.

## Appendix: Rebuilding vs. reusing the same binary

PGO introduces non-determinacy in our builds, so even builds built from the same source may have different performance. Given this, I thought it was a more "honest" A/A test to rebuild twice from the same source rather than reusing the binary, given that PGO remains an important thing to do.

**Comment by mdboom** [comment]
Another use of the A/A test here is to identify specific benchmarks with high variability (e.g. `spectralnorm`) and either improve them (maybe more repetitions or something) or remove them because they only introduce noise in the overall system.

**Comment by brandtbucher** [comment]
Thanks for collecting this data!

Do we see any improvement in the situation with pyperformance's `--rigorous` option? Even if it's not enough to stabilize the CI runs, maybe it's worth turning on on our benchmarking machine (I'm not sure if we currently do).

Also, the `pidigits` "step function" results are something that we've observed on dedicated hardware before - it's not unusual to see exact +/-11% swings when building the *exact same commit* (see the expanded details in [this comment](https://github.com/faster-cpython/ideas/discussions/142#discussion-3719108)). We think it's just an artifact of the benchmark.

**Comment by mdboom** [comment]
> Do we see any improvement in the situation with pyperformance's `--rigorous` option? Even if it's not enough to stabilize the CI runs, maybe it's worth turning on on our benchmarking machine (I'm not sure if we currently do).

Good suggestion -- I didn't experiment with that.

> Also, the pidigits "step function" results are something that we've observed on dedicated hardware before - it's not unusual to see exact +/-11% swings when building the exact same commit (see the expanded details in https://github.com/faster-cpython/ideas/discussions/142#discussion-3719108). We think it's just an artifact of the benchmark.

Oh interesting.  It's not on these specific runs on bare metal, so I assumed it was an effect of the cloud machine.  That just sort of underlines that compensating for the effects of running in the cloud is even harder than I already thought.

**Comment by ericsnowcurrently** [comment]
Are we able to pin CPUs on the cloud runs, like we do on bare-metal?  IIRC, that had a meaningful effect on the stability of results.

Also, for cloud results, do benchmarks run in parallel on the same VM?  If so, I'd expect the parallel processes to impact each other's results, skewing the individual results, as well as the overall variability between runs.  If not run on the same VM, what is the effect on the aggregate results due to the possible variability?

**Comment by mdboom** [comment]
> Are we able to pin CPUs on the cloud runs, like we do on bare-metal? IIRC, that had a meaningful effect on the stability of results.

Yes, we're running the same pinning on the cloud runs -- I didn't confirm whether or not that has any effect.

> Also, for cloud results, do benchmarks run in parallel on the same VM? If so, I'd expect the parallel processes to impact each other's results, skewing the individual results, as well as the overall variability between runs. If not run on the same VM, what is the effect on the aggregate results due to the possible variability?

I think it's likely they are running on the same physical machine, but I don't know if there is any way to know.  We could not run them in parallel, which would prevent *our* benchmarks from running at the same time, but wouldn't prevent other work from happening on the same machine.

**Comment by mdboom** [comment]
In order to separate out the effect of PGO on variability of the results, I did another A/A test on bare metal, but using the *exact same binary* both times, rather than a different build from the same source (where PGO and layout randomization is likely to introduce variation).

**tl;dr** It doesn't seem to matter much for variability.

![plot](https://user-images.githubusercontent.com/38294/196967987-bf694033-0f6c-44b5-97f4-bc5d9f7d1f73.png)

**Different builds:**

![quantiles-different](https://user-images.githubusercontent.com/38294/196968105-07397dc1-41da-4222-9041-b657be162b80.png)

**Same builds:**

![quantiles-same](https://user-images.githubusercontent.com/38294/196968171-ff0a4ee7-58c3-4376-a74d-49dc8192378c.png)

**Comment by jakebailey** [comment]
I just wanted to note my thoughts here (from the TypeScript x Python meeting chat) as I think they're somewhat interesting and maybe you can come up with something too.

I am dealing with what appears to be the same problem; I am using `pyperf system tune` and core isolation to make benchmarks more stable (thanks for sharing the links!). But, I would run the benchmarks one day, come back and rerun them the next day, and my results differ enough that normal ways of checking for changes (a U test) says that something changed, which is bad. Even though everything I can think of is controlled, something about the system is behaving differently. (temperature? my mood? something else?)

One method I have found to help is to use a "proxy benchmark"; before each run, I do some CPU bound task (like a big busy loop, SHA512-ing random data, etc), which gives a datapoint as to how the machine is performing "right now". Then, I can normalize each of the real benchmarks by dividing by its proxy benchmark (which "feels" better if the proxy is a convenient number like 1 second). When I compare the normalized benchmarks, my A/A test tends to no longer say that my results have changed.

This really only works if the benchmark is CPU bound, and it's hard to say if plain division is the right way to combine the two benchmarks, but it appears to be better than nothing.

**Comment by ericsnowcurrently** [comment]
CC @vstinner

**Comment by mkbosmans** [comment]
>  Anecdotally, this seems to be worse on benchmarks that perform I/O (pathlib).
It might be useful here to load all files required for testing from a tmpfs. That will reduce a lot of the I/O time and the variability the causes.

Of course this would not allow you to benchmark performance improvements that target I/O operations in a realistic way, so some care needs to be taken when interpreting results.

**Comment by gvanrossum** [comment]
A new discussion may appear on Discord:
https://discuss.python.org/t/python-benchmarking-in-unstable-environments/22334

**Comment by jakebailey** [comment]
An update on my above "proxy benchmark" suggestion a few months on: I tested that method out but I'm sad to say that I ended up ripping it out entirely. The TypeScript benchmarks were just too long for the idea to work (if it was even sound in the first place, which I suspect it is not).

I have however now added a Mann-Whitney U-Test to our benchmark comparison infrastructure (which is what Go's benchstat tool uses), which now allows us to focus on benchmarking results that actually are within significant bounds by hiding any %-change that isn't p<0.05. The main problem we have is noise, so this is another (pretty standard) way to prevent making decisions based on said noise.
